fix: replace bad rid panics with errors

bartlomieju opened the pull request:
Closes #2249
Closes #2406

bartlomieju commented on cli/resources.rs:
```diff
@@ -101,6 +101,13 @@ enum Repr {
+    ErrorKind::NotFound,
+    format!("Resource with rid {:?} not found", rid),
+  )
+}
```
Thanks, looks like [`from_raw_os_error`](https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error) is the tool for the job, because there's no error kind like that in `std::io::ErrorKind`.

ry replied:
Use BadResource

https://github.com/denoland/deno/blob/ca000392857b4c79a3609ddbc20073222498998b/cli/msg.rs#L28

https://github.com/denoland/deno/blob/ca000392857b4c79a3609ddbc20073222498998b/cli/deno_error.rs#L49-L51

bartlomieju replied:
@ry that's not possible... We need this error to be of `std::io::Error` type because `AsyncWrite` trait needs that, see https://github.com/denoland/deno/issues/2249 for longer description

bartlomieju commented:
@ry @piscisaureus should we discriminate situation when resource was not found and doesn't match required repr?

eg:
```rust
impl AsyncWrite for Resource {
  fn poll_write(&mut self, buf: &[u8]) -> Poll<usize, Error> {
    let mut table = RESOURCE_TABLE.lock().unwrap();
    
    // if resource not found it returns OS error 9 (EBADF)
    let repr = table.get_mut(&self.rid).ok_or_else(resource_not_found)?;

    match repr {
      Repr::FsFile(ref mut f) => f.poll_write(buf),
      Repr::Stdout(ref mut f) => f.poll_write(buf),
      Repr::Stderr(ref mut f) => f.poll_write(buf),
      Repr::TcpStream(ref mut f) => f.poll_write(buf),
      Repr::ChildStdin(ref mut f) => f.poll_write(buf),
      // custom error (ErrorKind === 19), maybe rename to bad_resource_kind?
      r => bad_resource();
    }
  }
}
```

bartlomieju commented on tests/044_bad_resource.ts:
```diff
@@ -0,0 +1,7 @@
+async function main(): Promise<void> {
```
Test that is failing on master:
```
async function main(): Promise<void> {
  const file = await Deno.open("Cargo.toml", "a");
  file.write(new Uint8Array([0, 1, 2]));
  file.close();
}

main();

```

Unfortunately minimal dispatch still swallows errors and it can't be used right now (because it'll throw generic write error instead of displaying actual error)

ry replied:
Yeah, minimal_dispatch doesn't have error handling... I need to get to that.

bartlomieju commented:
After some thought I'm not sure that returning `Uncaught Other: Bad file descriptor (os error 9)` is a proper way to handle missing resources. Resources are not "file descriptors" per se, they are abstraction over file descriptors.

How about those:
- missing resource in resource table:
```rust
std::io::Error(
    std::io::ErrorKind::Other, 
    "Bad resource descriptor <rid>"
)
```
- mismatch of resource kind for operation (eg. trying to shutdown a file instead of TcpStream):
```rust
std::io::Error(
    std::io::ErrorKind::Other, 
    "Bad resource kind <rid> (<rid_repr>)"
)
```

ry commented on cli/ops/files.rs:
```diff
@@ -129,17 +124,13 @@ pub fn op_seek(
-        Ok(JsonOp::Async(Box::new(op)))
-      }
-    }
+  let resource = resources::lookup(args.rid as u32)?;
```
If it's a bad rid, you should get a BadResource error from `lookup()`

bartlomieju replied:
Alright, but then again inside `poll_read` and `poll_write` if `rid` is not present in resource table we'd return different error type (`std::io::Error` because `AsyncRead/Write` requires it) than during lookup.

ry commented on cli/ops/files.rs:
```diff
@@ -129,17 +124,13 @@ pub fn op_seek(
-      }
-    }
+  let resource = resources::lookup(args.rid as u32)?;
+  let op = resources::seek(resource, args.offset, args.whence as u32)
```
Therefore it's impossible to call resources::seek with a bad resource.

ry commented on cli/resources.rs:
```diff
@@ -239,18 +249,18 @@ impl Read for Resource {
     let mut table = RESOURCE_TABLE.lock().unwrap();
-    let maybe_repr = table.get_mut(&self.rid);
-    match maybe_repr {
-      None => panic!("bad rid"),
```
I think the panic is okay here. See above comments about `lookup()`.

bartlomieju replied:
This panic is exactly the problem we're trying to fix. `poll_read` is called after lookup, but since it's a future it can be called after the resource is removed (see integration test I provided)

ry replied:
Oh ok - sorry - I hadn't understood that. The problem is that we're implementing AsyncRead and that only returns std::io::Errors... I wonder if we can somehow to this without using AsyncRead, and return DenoError.

ry commented on cli/deno_error.rs:
```diff
@@ -46,10 +46,6 @@ impl fmt::Display for StaticError {
 
-pub fn bad_resource() -> ErrBox {
-  StaticError(ErrorKind::BadResource, "bad resource id").into()
-}
```
I don't agree this is wrong...

ry commented on cli/ops/files.rs:
```diff
@@ -104,13 +103,9 @@ pub fn op_close(
-  }
+  let resource = resources::lookup(args.rid as u32)?;
+  resource.close();
+  Ok(JsonOp::Sync(json!({})))
```
👍 
This seems good.

ry commented on cli/resources.rs:
```diff
@@ -101,6 +100,17 @@ enum Repr {
 
+fn bad_resource(rid: ResourceId) -> Error {
+  Error::new(ErrorKind::Other, format!("Bad resource descriptor {}", rid))
+}
```
Use ErrorKind::BadResource

bartlomieju commented:
@ry I implemented `DenoAsyncRead` and `DenoAsyncWrite`, changed the signature of `tokio_write::write` and added `tokio_read::read` as well as reverted back to `deno_error::bad_resources` - the only difference between them and regular tokio implementations is that it uses `ErrBox` instead of `std::io::Error`. 

PTAL and let me know what you think, it's not that much of the code and definitely gives better error handling. If you're good with that we can override traits for accept et al

bartlomieju commented on cli/deno_error.rs:
```diff
@@ -46,10 +46,14 @@ impl fmt::Display for StaticError {
 }
 
-pub fn bad_resource() -> ErrBox {
+pub fn bad_resource(_rid: u32) -> ErrBox {
```
It'd be cool if we could annotate the error message with actual `rid` but for now I'm getting errors:
```
error[E0716]: temporary value dropped while borrowed
  --> cli/deno_error.rs:54:40
   |
54 |   StaticError(ErrorKind::BadResource, &format!("bad resource kind {:?}", rid.clone())).into()
   |                                       -^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                                       ||
   |                                       |creates a temporary which is freed while still in use
   |                                       requires that borrow lasts for `'static`
55 | }
   | - temporary value is freed at the end of this statement
```

This can be figured out in separate PR tho

ry replied:
It's probably because `StaticError` requires a `&'static str`. I think there are other ways to construct an ErrBox which don't require a `&'static str` error message. But I agree that this can be done later.

ry commented on cli/deno_error.rs:
```diff
@@ -50,6 +50,10 @@ pub fn bad_resource() -> ErrBox {
 }
 
+pub fn bad_resource_kind() -> ErrBox {
+  StaticError(ErrorKind::BadResource, "bad resource kind").into()
```
"bad resource kind" is not a very descriptive message - it took me a few minutes to figure out the difference from "bad resource id". 

I question the value of distinguishing between these two types of errors... What do you think about replacing this with the existing `bad_resource()` error?

bartlomieju replied:
Yeah it makes sense :+1: I'll replace them

ry commented on cli/resources.rs:
```diff
@@ -236,22 +235,28 @@ impl Read for Resource {
 
-impl AsyncRead for Resource {
-  fn poll_read(&mut self, buf: &mut [u8]) -> Poll<usize, Error> {
+pub trait DenoAsyncRead {
```
Could you add a triple slash comment here describing the purpose of DenoAsyncRead (same for DenoAsyncWrite)

Something like this: /// DenoAsyncRead is the same as the AsyncRead trait but uses an ErrBox error instead of std::io:Error

ry commented on cli/tokio_write.rs:
```diff
@@ -29,23 +28,25 @@ enum State<A, T> {
 }
 
+/// This is almost the same implementation as in tokio, the only difference is
+/// that error type is `ErrBox` instead of `std::io::Error`.
```
👍

ry commented:
Just a few comments - otherwise looks good

bartlomieju commented:
Waiting on #2896

bartlomieju commented:
Rebased on top of `master`
